rllib/agents/dreamer/dreamer.py

```diff
@@ -0,0 +1,254 @@
+        """
+        episodes_buffer = []
+        while len(episodes_buffer) < batch_size:
+            rand_index = random.randint(0, len(self.episodes) - 1)
+            episode = self.episodes[rand_index]
+            if episode.count < self.length:
+                continue
+            available = episode.count - self.length
+            index = int(random.randint(0, available))
+            episodes_buffer.append(episode.slice(index, index + self.length))
+
+        batch = {}
+        for k in episodes_buffer[0].keys():
+            batch[k] = np.stack([e[k] for e in episodes_buffer], axis=0)
+
+        return SampleBatch(batch)
+
+
+def total_sampled_timesteps(worker):
+    return worker.policy_map[DEFAULT_POLICY_ID].global_timestep
+
+
+def execution_plan(workers, config):
+    # Special Replay Buffer for Dreamer agent
+    episode_buffer = EpisodicBuffer(length=config["batch_length"])
+
+    local_worker = workers.local_worker()
+
+    # Prefill episode buffer with initial exploration (uniform sampling)
+    while total_sampled_timesteps(local_worker) < config["prefill_timesteps"]:
+        samples = local_worker.sample()
+        episode_buffer.add(samples)
+
+    batch_size = config["batch_size"]
+    dreamer_train_iters = config["dreamer_train_iters"]
+    act_repeat = config["action_repeat"]
+
+    def dreamer_iteration(itr):
+        def postprocess_gif(gif: np.ndarray):
+            gif = np.clip(255 * gif, 0, 255).astype(np.uint8)
+            B, T, C, H, W = gif.shape
+            frames = gif.transpose((1, 2, 3, 0, 4)).reshape((1, T, C, H,
+                                                             B * W))
+            return frames
+
+        def policy_stats(fetches):
+            return fetches["default_policy"]["learner_stats"]
+
+        for samples in itr:
+            # Dreamer Training Loop
+            for n in range(dreamer_train_iters):
+                batch = episode_buffer.sample(batch_size)
+                if n == dreamer_train_iters - 1:
+                    batch["log_gif"] = True
+                fetches = local_worker.learn_on_batch(batch)
+
+            # Custom Logging
+            policy_fetches = policy_stats(fetches)
+            if "log_gif" in policy_fetches:
+                gif = policy_fetches["log_gif"]
+                policy_fetches["log_gif"] = postprocess_gif(gif)
+
+            metrics = _get_shared_metrics()
+            metrics.info[LEARNER_INFO] = fetches
+            metrics.counters[
+                STEPS_SAMPLED_COUNTER] = episode_buffer.timesteps * act_repeat
+            res = collect_metrics(local_worker=local_worker)
+            res["info"] = metrics.info
+            res["info"].update(metrics.counters)
+            res["timesteps_total"] = metrics.counters[STEPS_SAMPLED_COUNTER]
+            yield res
+
+            episode_buffer.add(samples)
+
+    rollouts = ParallelRollouts(workers)
+    rollouts = rollouts.transform(dreamer_iteration)
```

Comment:
Please use `.for_each()` instead of `.transform()` when possible. It doesn't look like you have loop-dependent state so you should be able to use a stateless function.